File: src/main/java/gay/skitbet/jackiro/utils/JackiroEmbed.java
package gay.skitbet.jackiro.utils;

import gay.skitbet.jackiro.Jackiro;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.Accessors;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.MessageEmbed;

import java.awt.*;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.EnumMap;
import java.util.List;

@NoArgsConstructor
@Accessors(chain = true)
@Getter
@Setter
public class JackiroEmbed {
    private static final String ICON_URL = Jackiro.getInstance().getShardManager().getShards().get(0).getSelfUser().getAvatarUrl();

    public static MessageEmbed getNewGuildEmbed(Guild guild) {
        return new JackiroEmbed()
                .setColor(Color.CYAN)
                .setTitle("Hello, I'm Jackiro!")
                .setDescription("Think your for adding me to your server! I'm a general purpose discord bot that can help keeps things running smootly!")
                .addField("Getting Started", "To get started you can use \"/setup\" to configure me!", false)
                .addField("Features", "• Automoderation\n• Custom Commands\n• Server Utilities\n• And much more!", false)
                .setThumbnailUrl(ICON_URL)
                .build();
    }

    private String title;
    private String description;
    private String footerText;
    private String imageUrl;
    private String thumbnailUrl;
    private Color color = Color.BLUE;
    private String authorName;
    private String authorUrl;
    private String authorIconUrl;
    private List<MessageEmbed.Field> fields = new ArrayList<>();

    /**
     * Builds embed!
     * @return
     */
    public MessageEmbed build() {
        EmbedBuilder embed = createEmbedBasics();

        if (title != null) embed.setTitle(title);
        if (description != null) embed.setDescription(description);
        if (footerText != null) embed.setFooter(footerText);
        if (imageUrl != null) embed.setImage(imageUrl);
        if (thumbnailUrl != null) embed.setThumbnail(thumbnailUrl);
        if (color != null) embed.setColor(color);
        if (authorName != null) embed.setAuthor(authorName, authorUrl, authorIconUrl);
        for (var field : fields) {
            embed.addField(field);
        }

        return embed.build();
    }

    public JackiroEmbed addField(String name, String value, boolean inLine) {
        fields.add(new MessageEmbed.Field(name, value, inLine));
        return this;
    }

    /**
     * creates a basic general error embed
     * @param message
     * @return
     */
    public MessageEmbed error(String message) {
        EmbedBuilder embed = createEmbedBasics();

        embed.setColor(Color.RED);
        embed.setTitle("Uh oh!");
        embed.setDescription("❌ An error has occured! \n" + message);
        return embed.build();
    }

    private EmbedBuilder createEmbedBasics() {
        EmbedBuilder embed = new EmbedBuilder();

        embed.setTimestamp(LocalDateTime.now());
        embed.setFooter("Jackiro", ICON_URL);

        return embed;
    }

}

File: src/main/kotlin/gay/skitbet/jackiro/utils/JackiroEmbed.kt
package gay.skitbet.jackiro.utils

import gay.skitbet.jackiro.Jackiro
import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.entities.MessageEmbed
import java.awt.Color
import java.time.LocalDateTime

class JackiroEmbed {
    var title: String? = null
    var description: String? = null
    var footerText: String? = null
    var imageUrl: String? = null
    var thumbnailUrl: String? = null
    var color: Color? = Color.BLUE
    var authorName: String? = null
    var authorUrl: String? = null
    var authorIconUrl: String? = null
    private val fields: MutableList<MessageEmbed.Field> = mutableListOf()

    fun setTitle(value: String?): JackiroEmbed = apply { title = value }
    fun setDescription(value: String?): JackiroEmbed = apply { description = value }
    fun setFooterText(value: String?): JackiroEmbed = apply { footerText = value }
    fun setImageUrl(value: String?): JackiroEmbed = apply { imageUrl = value }
    fun setThumbnailUrl(value: String?): JackiroEmbed = apply { thumbnailUrl = value }
    fun setColor(value: Color?): JackiroEmbed = apply { color = value }
    fun setAuthorName(value: String?): JackiroEmbed = apply { authorName = value }
    fun setAuthorUrl(value: String?): JackiroEmbed = apply { authorUrl = value }
    fun setAuthorIconUrl(value: String?): JackiroEmbed = apply { authorIconUrl = value }

    fun addField(name: String, value: String, inLine: Boolean): JackiroEmbed = apply {
        fields.add(MessageEmbed.Field(name, value, inLine))
    }

    fun build(): MessageEmbed {
        val embed = createEmbedBasics()

        title?.let { embed.setTitle(it) }
        description?.let { embed.setDescription(it) }
        footerText?.let { embed.setFooter(it) }
        imageUrl?.let { embed.setImage(it) }
        thumbnailUrl?.let { embed.setThumbnail(it) }
        color?.let { embed.setColor(it) }

        if (authorName != null) {
            embed.setAuthor(authorName, authorUrl, authorIconUrl)
        }

        fields.forEach { embed.addField(it) }

        return embed.build()
    }

    fun error(message: String): MessageEmbed {
        return EmbedBuilder()
            .setTimestamp(LocalDateTime.now())
            .setColor(Color.RED)
            .setFooter("Jackiro", ICON_URL)
            .setTitle("Uh oh!")
            .setDescription("❌ An error has occurred! \n$message")
            .build()
    }

    private fun createEmbedBasics(): EmbedBuilder {
        return EmbedBuilder().apply {
            setTimestamp(LocalDateTime.now())
            setFooter("Jackiro", ICON_URL)
        }
    }

    companion object {
        private val ICON_URL: String? = Jackiro.instance
            .shardManager
            .shards
            .firstOrNull()
            ?.selfUser
            ?.avatarUrl

        fun getNewGuildEmbed(guild: Guild): MessageEmbed {
            return JackiroEmbed()
                .setColor(Color.CYAN)
                .setTitle("Hello, I'm Jackiro!")
                .setDescription("Thanks for adding me to your server! I'm a general-purpose Discord bot that helps keep things running smoothly!")
                .addField("Getting Started", "To get started you can use \"/setup\" to configure me!", false)
                .addField("Features", "• Automoderation\n• Custom Commands\n• Server Utilities\n• And much more!", false)
                .setThumbnailUrl(ICON_URL)
                .build()
        }
    }
}
